File: android/app/src/main/java/com/nauhaan/skycast/ui/home/HomeScreen.kt
package com.nauhaan.skycast.ui.home

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.pager.HorizontalPager
import androidx.compose.foundation.pager.rememberPagerState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.List
import androidx.compose.material.icons.filled.AddLocationAlt
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nauhaan.skycast.R
import com.nauhaan.skycast.core.designsystem.component.EmptyStateView
import com.nauhaan.skycast.core.designsystem.component.ErrorView
import com.nauhaan.skycast.core.designsystem.component.LoadingView
import com.nauhaan.skycast.core.designsystem.component.PageScrubber
import com.nauhaan.skycast.core.designsystem.component.StaleDataBanner
import com.nauhaan.skycast.core.designsystem.component.WeatherBackground
import com.nauhaan.skycast.core.designsystem.component.WeatherDetailGrid
import com.nauhaan.skycast.core.designsystem.theme.SkyCastTheme
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.core.designsystem.theme.weatherSurfaceTint
import com.nauhaan.skycast.domain.model.SavedLocation
import com.nauhaan.skycast.domain.model.WeatherCondition
import com.nauhaan.skycast.domain.usecase.TodayLocationWeather
import com.nauhaan.skycast.ui.common.CurrentConditionsHeader
import com.nauhaan.skycast.ui.common.rememberWeatherDetailLabels
import com.nauhaan.skycast.ui.common.toDetails
import com.nauhaan.skycast.ui.common.toPresentation

/**
 * The Home tab.
 *
 * - [HomeScreen] is the **stateful** entry point. It obtains the view model and does nothing else.
 * - [HomeContent] is **stateless**: state in, lambdas out, so it is previewable in every state and
 *   assertable in a Compose test without Hilt.
 */
@Composable
fun HomeScreen(
    onNavigateToLocationDetail: (Long) -> Unit,
    onNavigateToAddLocation: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: HomeViewModel = hiltViewModel(),
    onWeatherTintChanged: (Color?) -> Unit = {},
) {
    // collectAsStateWithLifecycle, not collectAsState: stops collection when the
    // screen is not visible instead of doing work in the background.
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    // Reported upwards so the navigation bar can share this page's mood. The bar is a sibling of
    // the content rather than a descendant, so the CompositionLocal the tiles read cannot reach
    // it, state goes up by lambda and the colour comes back down as a parameter.
    val weather = uiState.weather
    val tint = weather?.let { weatherSurfaceTint(it.condition, it.isDaytime) }
    LaunchedEffect(tint) { onWeatherTintChanged(tint) }

    HomeContent(
        uiState = uiState,
        onRefresh = viewModel::refresh,
        onDismissBanner = viewModel::dismissBanner,
        onSelectPage = viewModel::selectPage,
        onOpenDetail = onNavigateToLocationDetail,
        onAddLocation = onNavigateToAddLocation,
        modifier = modifier,
    )
}

@Composable
internal fun HomeContent(
    uiState: HomeUiState,
    onRefresh: () -> Unit,
    onDismissBanner: () -> Unit,
    onSelectPage: (Int) -> Unit,
    onOpenDetail: (Long) -> Unit,
    onAddLocation: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // Exactly one branch renders. The ordering encodes the offline-first rules from
    // The offline-first read algorithm: content wins over errors whenever a cache exists.
    when {
        uiState.showsFullScreenLoader ->
            LoadingView(modifier = modifier, message = stringResource(R.string.home_loading))

        uiState.showsEmptyState ->
            EmptyStateView(
                title = stringResource(R.string.home_empty_title),
                message = stringResource(R.string.home_empty_message),
                icon = Icons.Filled.AddLocationAlt,
                actionLabel = stringResource(R.string.action_add_location),
                onAction = onAddLocation,
                modifier = modifier,
            )

        uiState.showsFullScreenError -> {
            val presentation = requireNotNull(uiState.error).toPresentation()
            ErrorView(
                title = stringResource(presentation.titleRes),
                message = stringResource(presentation.messageRes),
                icon = Icons.Filled.CloudOff,
                onRetry = onRefresh.takeIf { presentation.isRetryable },
                modifier = modifier,
            )
        }

        else -> HomePager(
            uiState = uiState,
            onRefresh = onRefresh,
            onDismissBanner = onDismissBanner,
            onSelectPage = onSelectPage,
            onOpenDetail = onOpenDetail,
            modifier = modifier,
        )
    }
}

/**
 * The saved places, one per page.
 *
 * A pager rather than a switcher alone: swiping between places is how weather apps on both
 * platforms work, and the gesture feels immediate in a way a menu does not. The menu above it
 * stays, because a gesture with no visible affordance is undiscoverable in the other direction,
 * someone who never swipes would never learn there is more than one page.
 *
 * Every page's data is already in memory (see `ObserveTodayWeatherUseCase`), so a swipe reveals a
 * loaded screen rather than a spinner.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun HomePager(
    uiState: HomeUiState,
    onRefresh: () -> Unit,
    onDismissBanner: () -> Unit,
    onSelectPage: (Int) -> Unit,
    onOpenDetail: (Long) -> Unit,
    modifier: Modifier = Modifier,
) {
    val pagerState = rememberPagerState(
        initialPage = uiState.selectedIndex,
        pageCount = { uiState.pages.size },
    )

    // Two-way binding between the pager and the view model: a swipe reports the new page, and
    // choosing from the menu animates the pager to it. `snapshotFlow` on `settledPage` rather than
    // a callback, so the report happens once the page has come to rest instead of on every frame
    // of the drag.
    LaunchedEffect(pagerState) {
        snapshotFlow { pagerState.settledPage }.collect(onSelectPage)
    }
    LaunchedEffect(uiState.selectedIndex) {
        if (uiState.selectedIndex != pagerState.currentPage) {
            pagerState.animateScrollToPage(uiState.selectedIndex)
        }
    }

    // The background follows the place on screen, so swiping from a clear Malé to an overcast
    // London visibly changes the weather of the whole screen, not just the numbers on it.
    val selected = uiState.selected?.weather?.data

    WeatherBackground(
        condition = selected?.condition ?: WeatherCondition.UNKNOWN,
        isDaytime = selected?.isDaytime ?: true,
        modifier = modifier.fillMaxSize(),
    ) {
        Column(modifier = Modifier.fillMaxSize()) {
            LocationMenu(
                uiState = uiState,
                onSelectPage = onSelectPage,
                modifier = Modifier
                    .align(Alignment.End)
                    .padding(horizontal = Spacing.sm, vertical = Spacing.xs),
            )

            PullToRefreshBox(
                isRefreshing = uiState.isRefreshing,
                onRefresh = onRefresh,
                modifier = Modifier.fillMaxSize(),
            ) {
                HorizontalPager(state = pagerState) { page ->
                    HomePage(
                        page = uiState.pages[page],
                        uiState = uiState,
                        isSelected = page == uiState.selectedIndex,
                        onRefresh = onRefresh,
                        onDismissBanner = onDismissBanner,
                        onOpenDetail = onOpenDetail,
                        onSelectPage = onSelectPage,
                    )
                }
            }
        }
    }
}

/** One page: the hero reading, the hourly strip and the detail tiles for a single place. */
@Composable
private fun HomePage(
    page: TodayLocationWeather,
    uiState: HomeUiState,
    isSelected: Boolean,
    onRefresh: () -> Unit,
    onDismissBanner: () -> Unit,
    onOpenDetail: (Long) -> Unit,
    onSelectPage: (Int) -> Unit,
    modifier: Modifier = Modifier,
) {
    val weather = page.weather.data

    Column(
        modifier = modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState()),
    ) {
        // The banner belongs to the page on screen: a stale reading for Malé should not warn the
        // user while they are looking at London.
        if (isSelected && uiState.showsStaleBanner) {
            val message = uiState.error
                ?.let { stringResource(it.toPresentation().messageRes) }
                ?: stringResource(R.string.banner_data_may_be_out_of_date)
            StaleDataBanner(message = message, onRetry = onRefresh, onDismiss = onDismissBanner)
        }

        if (weather == null) {
            // A page reached by swiping ahead of its data.
            LoadingView(message = stringResource(R.string.home_loading))
            return@Column
        }

        PlaceHeading(
            location = page.location,
            modifier = Modifier.padding(horizontal = Spacing.md),
        )

        CurrentConditionsHeader(
            weather = weather,
            unit = uiState.preferences.temperatureUnit,
            showsLocationName = false,
            // Tapping the hero block pushes the full detail screen, the push half of the
            // navigation hierarchy, reachable from Home.
            onClick = { onOpenDetail(weather.locationId) },
            modifier = Modifier.padding(Spacing.md),
        )

        // Between the reading and the strip, centred: the indicator belongs to the pager, so it
        // sits directly under what paging changes rather than off in a corner of the chrome.
        if (uiState.showsPageIndicator) {
            PageScrubber(
                count = uiState.pages.size,
                selectedIndex = uiState.selectedIndex,
                onSelect = onSelectPage,
                contentDescription = stringResource(
                    R.string.home_showing_location,
                    page.location.name,
                    uiState.selectedIndex + 1,
                    uiState.pages.size,
                ),
                modifier = Modifier
                    .align(Alignment.CenterHorizontally)
                    .padding(bottom = Spacing.sm),
            )
        }

        HourlyStrip(
            // Every page draws its own strip, including the ones off-screen. The slice is cheap,
            // and skipping it made the section appear mid-swipe.
            hours = uiState.hourlyWindow(page),
            zoneOffset = weather.zoneOffset,
            unit = uiState.preferences.temperatureUnit,
            modifier = Modifier.padding(bottom = Spacing.md),
        )

        WeatherDetailGrid(
            // Six tiles here, eight on the detail screen: Home is a glance, and the derived
            // readings are why someone taps through.
            details = weather.toDetails(
                preferences = uiState.preferences,
                labels = rememberWeatherDetailLabels(),
            ),
            modifier = Modifier.padding(horizontal = Spacing.md),
        )
    }
}

/**
 * A menu of the saved places, for jumping straight to one.
 *
 * Icon-only, and only present when there is more than one place.
 */
@Composable
private fun LocationMenu(uiState: HomeUiState, onSelectPage: (Int) -> Unit, modifier: Modifier = Modifier) {
    if (!uiState.showsPageIndicator) return
    var isMenuOpen by remember { mutableStateOf(false) }

    Box(modifier = modifier) {
        IconButton(onClick = { isMenuOpen = true }) {
            Icon(
                imageVector = Icons.AutoMirrored.Filled.List,
                contentDescription = stringResource(R.string.home_select_location),
            )
        }
        DropdownMenu(expanded = isMenuOpen, onDismissRequest = { isMenuOpen = false }) {
            uiState.pages.forEachIndexed { index, entry ->
                DropdownMenuItem(
                    text = { Text(entry.location.displayName) },
                    // A tick beside the place on screen; a column of identical rows cannot say
                    // which one you are looking at.
                    trailingIcon = {
                        if (index == uiState.selectedIndex) {
                            Icon(
                                imageVector = Icons.Filled.Check,
                                contentDescription = null,
                            )
                        }
                    },
                    onClick = {
                        onSelectPage(index)
                        isMenuOpen = false
                    },
                )
            }
        }
    }
}

/**
 * The place, as the page's heading.
 *
 * Large and at the top of the content, because it is the single most important word on the screen
 * and it spent a while as the smallest, a label inside a text button in the corner.
 *
 * The region line beneath it earns its space when two saved places share a name, which is common
 * enough (there are more than twenty Londons) that "London" alone can be genuinely ambiguous.
 */
@Composable
private fun PlaceHeading(location: SavedLocation, modifier: Modifier = Modifier) {
    val region = location.displayName
        .split(",")
        .drop(1)
        .joinToString(", ") { it.trim() }
        .takeIf { it.isNotBlank() }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                contentDescription = listOfNotNull(location.name, region).joinToString(". ")
            },
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(
            text = location.name,
            // The emphasized display role: this is the one place on Home that earns it, alongside
            // the temperature. Two per screen is the ceiling Material 3 sets.
            style = MaterialTheme.typography.headlineLargeEmphasized,
            textAlign = TextAlign.Center,
        )
        if (region != null) {
            Text(
                text = region,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
            )
        }
    }
}

@Preview(name = "Loading", showBackground = true)
@Composable
private fun TodayLoadingPreview() {
    SkyCastTheme {
        HomeContent(
            uiState = HomeUiState(isLoading = true),
            onRefresh = {},
            onDismissBanner = {},
            onSelectPage = {},
            onOpenDetail = {},
            onAddLocation = {},
        )
    }
}

@Preview(name = "Empty", showBackground = true)
@Composable
private fun TodayEmptyPreview() {
    SkyCastTheme {
        HomeContent(
            uiState = HomeUiState(hasNoLocation = true),
            onRefresh = {},
            onDismissBanner = {},
            onSelectPage = {},
            onOpenDetail = {},
            onAddLocation = {},
        )
    }
}

File: ios/SkyCast/Features/Home/HomeScreen.swift
import SwiftUI

/// The Home tab.
///
/// - ``HomeScreen`` is the **stateful** entry point. It builds the view model from the
///   environment container and does nothing else.
/// - ``HomeContent`` is **stateless**: state in, closures out.
struct HomeScreen: View {
    @Environment(AppContainer.self) private var container
    @State private var viewModel: HomeViewModel?
    /// The pushed day-detail destination. `nil` means nothing is pushed.
    @State private var selectedDay: DayRoute?

    var body: some View {
        Group {
            if let viewModel {
                HomeContent(
                    state: viewModel.state,
                    onRefresh: { await viewModel.refresh() },
                    onDismissBanner: viewModel.dismissBanner,
                    onSelectPage: viewModel.selectPage,
                    onSelectDay: { locationID, date in selectedDay = DayRoute(locationID: locationID, date: date) }
                )
            } else {
                LoadingView()
            }
        }
        .toolbarTitleDisplayMode(.inline)
        .navigationDestination(item: $selectedDay) { route in
            DayDetailScreen(locationID: route.locationID, date: route.date)
        }
        .task {
            // @Environment is not available during init, so the view model is built here.
            if viewModel == nil {
                viewModel = HomeViewModel(
                    weatherRepository: container.weatherRepository,
                    locationRepository: container.locationRepository,
                    settingsStore: container.settingsStore,
                    selectedLocationStore: container.selectedLocationStore
                )
                viewModel?.start()
            } else {
                // A place may have been added or removed on the Locations tab while this screen
                // was off-screen.
                await viewModel?.reload()
            }
        }
        .onDisappear { viewModel?.stop() }
    }
}

/// One day of one place, which is what Home has to carry to push the day-detail screen.
///
/// Home pages across several places, so a bare `Date` would be ambiguous: tapping Thursday on Malé
/// must not open Thursday on London.
struct DayRoute: Hashable {
    let locationID: Int64
    let date: Date
}

/// The stateless half.
struct HomeContent: View {
    /// Measured from the sticky header so each page can start below it. A constant would break
    /// when the pill appears or disappears, or when Dynamic Type resizes the name.
    @State private var headerHeight: CGFloat = 0

    let state: HomeUiState
    let onRefresh: () async -> Void
    let onDismissBanner: () -> Void
    let onSelectPage: (Int) -> Void
    let onSelectDay: (Int64, Date) -> Void

    var body: some View {
        // Exactly one branch renders. The ordering encodes the offline-first rules from
        // The offline-first read algorithm: content wins over errors whenever a cache exists.
        if state.showsFullScreenLoader {
            LoadingView(message: "Fetching the latest weather…")
        } else if state.showsEmptyState {
            EmptyStateView(
                title: "No locations yet",
                message: "Add a place and SkyCast will keep its forecast ready, even offline.",
                systemImage: "mappin.and.ellipse"
            )
        } else if state.showsFullScreenError, let error = state.error {
            ErrorView(error: error, onRetry: { Task { await onRefresh() } })
        } else {
            pager
        }
    }

    /// The saved places, one per page.
    ///
    /// `TabView` in page style supplies the gesture, the rubber-banding, the index binding and
    /// VoiceOver's swipe navigation between pages. The place name lives in the toolbar, and the
    /// menu beside it offers the list for readers who do not swipe.
    private var pager: some View {
        // The GeometryReader recovers the safe-area insets after they are ignored. A paged
        // TabView lays its pages out inside the safe area, so the pager ignores the insets and each
        // page puts them back as content padding.
        GeometryReader { proxy in
            TabView(selection: selectionBinding) {
                ForEach(Array(state.pages.enumerated()), id: \.element.id) { index, page in
                    HomePageView(
                        page: page,
                        state: state,
                        isSelected: index == state.selectedIndex,
                        onRefresh: onRefresh,
                        onDismissBanner: onDismissBanner,
                        onSelectDay: onSelectDay,
                        onSelectPage: onSelectPage,
                        chromeInsets: proxy.safeAreaInsets,
                        headerHeight: headerHeight
                    )
                    .tag(index)
                }
            }
            .tabViewStyle(.page(indexDisplayMode: .never))
            // On the TabView, not on the GeometryReader: a reader that ignores the safe area
            // reports no insets, which would leave the pages with no padding.
            .ignoresSafeArea(.container, edges: .vertical)
        }
        // Pinned above the pages, so the place name stays legible while a page scrolls under it
        // and the dots keep a fixed position.
        //
        // Attached outside the `GeometryReader`, whose frame ignores the safe area and so gives
        // nothing sensible to align to.
        .overlay(alignment: .top) {
            HomeStickyHeader(state: state, onSelectPage: onSelectPage)
                .onGeometryChange(for: CGFloat.self) { $0.size.height } action: { headerHeight = $0 }
        }
        // Without this the navigation bar paints its own background across the top of the screen.
        // Hiding it leaves only the toolbar items.
        .toolbarBackgroundVisibility(.hidden, for: .navigationBar)
        // Follows the place on screen, so swiping from a clear Malé to an overcast London changes
        // the weather of the whole screen, not just the numbers on it.
        .weatherBackground(
            condition: state.weather?.condition ?? .unknown,
            isDaytime: state.weather?.isDaytime ?? true
        )
    }

    private var selectionBinding: Binding<Int> {
        Binding(get: { state.selectedIndex }, set: onSelectPage)
    }
}

/// One page: the hero reading, the hourly strip and the detail tiles for a single place.
private struct HomePageView: View {
    let page: HomePage
    let state: HomeUiState
    let isSelected: Bool
    let onRefresh: () async -> Void
    let onDismissBanner: () -> Void
    let onSelectDay: (Int64, Date) -> Void
    let onSelectPage: (Int) -> Void
    /// The insets the pager gave up so content could run behind the bars; see
    /// ``HomeContent/pager``. Re-applied here as content padding.
    let chromeInsets: EdgeInsets
    /// How far down the page's content has to start to clear the pinned header.
    let headerHeight: CGFloat

    var body: some View {
        ScrollView {
            // One glass container for every floating surface on this page, so the banner and hero
            // behave as a single material.
            SkyGlassGroup(spacing: Spacing.md) {
                VStack(spacing: Spacing.md) {
                    // The banner belongs to the page on screen: a stale reading for Malé should
                    // not warn the user while they are looking at London.
                    if isSelected, state.showsStaleBanner {
                        StaleDataBanner(
                            message: state.staleBannerMessage,
                            onRetry: { Task { await onRefresh() } },
                            onDismiss: onDismissBanner
                        )
                    }

                    if let weather = page.weather.data {
                        CurrentConditionsHero(
                            weather: weather,
                            unit: state.preferences.temperatureUnit,
                            // The switcher above already names the place.
                            showsLocationName: false
                        )

                        HourlyStrip(
                            // Every page draws its own strip, including the ones off-screen. The
                            // slice is cheap, and skipping it made the section appear mid-swipe.
                            hours: state.hourlyWindow(for: page),
                            timeZone: weather.timeZone,
                            unit: state.preferences.temperatureUnit
                        )

                        // The forecast picture.
                        if let forecast = page.forecast.data {
                            TemperatureTrendSection(
                                forecast: forecast,
                                unit: state.preferences.temperatureUnit
                            )
                            DailyRangesSection(
                                forecast: forecast,
                                unit: state.preferences.temperatureUnit,
                                onSelectDay: { date in onSelectDay(weather.locationID, date) }
                            )
                        }

                        SectionHeader("Conditions")

                        if let sun = weather.sunPath() {
                            SunPathCard(
                                progress: sun.progress,
                                sunriseLabel: sun.sunriseLabel,
                                sunsetLabel: sun.sunsetLabel,
                                daylightLabel: sun.daylightLabel,
                                announcement: sun.announcement
                            )
                        }

                        // Absent at latitudes and dates with no such window; see `SolarCalculator`.
                        if let golden = goldenHourReading(
                            for: page.location,
                            timeZone: weather.timeZone
                        ) {
                            GoldenHourCard(reading: golden)
                        }

                        // All eight tiles, including the derived dew point and length of day. Home used
                        // to show six and keep those two behind the hero tap; with nowhere to tap
                        // through to, holding them back would just be hiding them.
                        WeatherDetailGrid(
                            details: weather.details(
                                preferences: state.preferences,
                                includeDerived: true
                            )
                        )

                        ObservedAtFooter(observedAt: weather.observedAt)
                    } else {
                        // A page reached by swiping ahead of its data.
                        LoadingView(message: "Fetching the latest weather…")
                    }
                }
                .padding(.horizontal, Spacing.md)
                // The header sits inside the safe area, so a page has to clear both it and the
                // status bar it passes behind.
                .padding(.top, chromeInsets.top + headerHeight)
                // The tab bar's own inset, plus room to scroll the last tile clear of it.
                .padding(.bottom, chromeInsets.bottom + Self.tabBarClearance)
            }
        }
        // Softens where content meets the navigation bar and the minimised tab bar.
        .scrollEdgeEffectStyle(.soft, for: .all)
        // Native pull-to-refresh; also drives the Retry action above.
        .refreshable { await onRefresh() }
    }

    /// Wide enough for the dots and no wider.
    ///
    /// Each dot occupies roughly 16 points, plus its own end padding.
    private func scrubberWidth(for count: Int) -> CGFloat {
        CGFloat(count) * Self.scrubberDotSpacing
    }

    private static let scrubberDotSpacing: CGFloat = 18
    private static let tabBarClearance: CGFloat = 56

    /// `UIPageControl` sizes itself; this only reserves the row.
    private static let scrubberHeight: CGFloat = 28
}

/// A menu of the saved places, for jumping straight to one.
///
/// Icon-only, and only present when there is more than one place. Shown as a `Picker` so the place
/// on screen carries a tick.
struct LocationMenu: View {
    let state: HomeUiState
    let onSelectPage: (Int) -> Void

    var body: some View {
        Menu {
            Picker("Place", selection: Binding(get: { state.selectedIndex }, set: onSelectPage)) {
                ForEach(Array(state.pages.enumerated()), id: \.element.id) { index, page in
                    Text(page.location.displayName).tag(index)
                }
            }
        } label: {
            Image(systemName: "list.bullet")
        }
        .accessibilityLabel("Choose a place, showing \(state.location?.name ?? "")")
    }
}

/// The place, as the page's heading.
///
/// The region line beneath it disambiguates saved places that share a name.
private struct PlaceHeading: View {
    let location: SavedLocation

    private var region: String? {
        let parts = location.displayName
            .split(separator: ",")
            .dropFirst()
            .map { $0.trimmingCharacters(in: .whitespaces) }
        return parts.isEmpty ? nil : parts.joined(separator: ", ")
    }

    var body: some View {
        VStack(spacing: Spacing.xxs) {
            Text(location.name)
                .font(.largeTitle.weight(.semibold))
                .minimumScaleFactor(0.6)
                .lineLimit(2)
                .multilineTextAlignment(.center)
            if let region {
                Text(region)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }
        }
        .frame(maxWidth: .infinity)
        // One announcement: a place and where it is, not two fragments.
        .accessibilityElement(children: .combine)
    }
}

// MARK: - Previews
// One per state, so every branch is reviewable without a device.

#Preview("Loading") {
    NavigationStack {
        HomeContent(
            state: HomeUiState(isLoading: true),
            onRefresh: {},
            onDismissBanner: {},
            onSelectPage: { _ in },
            onSelectDay: { _, _ in }
        )
    }
}

#Preview("Empty") {
    NavigationStack {
        HomeContent(
            state: HomeUiState(hasNoLocation: true),
            onRefresh: {},
            onDismissBanner: {},
            onSelectPage: { _ in },
            onSelectDay: { _, _ in }
        )
    }
}

#Preview("Offline, no cache") {
    NavigationStack {
        HomeContent(
            state: HomeUiState(refreshError: .offline),
            onRefresh: {},
            onDismissBanner: {},
            onSelectPage: { _ in },
            onSelectDay: { _, _ in }
        )
    }
}
